specification/hugr.md

```diff
@@ -1727,10 +1727,19 @@ allocation, and the compiler can add/remove/move these operations to use
 more or fewer qubits. In some use cases, that may not be desirable, and
 we may instead want to guarantee only a certain number of qubits are
 used by the program. For this purpose TKET2 places additional
-constraints on the `main` function that are in line with TKET1 backwards
-compatibility. Namely the main function takes one `Array<N, Qubit>`
+constraints on the HUGR that are in line with TKET1 backwards
+compatibility:
+
+1. The `main` function takes one `Array<N, Qubit>`
 input and has one output of the same type (the same statically known
-size). If further the program does not contain any `qalloc` or `qfree`
+size).
+2. All Operations that have a `FunctionType` involving `Qubit`:
+    1. Have as many `Qubit` input wires as output.
+    2. Guarantee semantics that when filtered to `Qubit` only, there is a
+       bijection such that the _nth_ input maps to the _nth_ output.
```

Comment:
It's a little unclear what this means.

"There is a bijection such that the _nth_ input maps to the _nth_ output" is trivially true if there are the same number of inputs and outputs.

Actually I had a look at the original slack thread and I am still not quite sure what it is trying to say. We are saying that the nth input "is" the nth output in some sense, but how to make this precise without introducing explicit qubit references (and saying that the nth input has the same qubit reference as the nth output)?

Reply:
To put it another way: what would failure of this condition look like?

Reply:
Yeah I am not sure how to express it either. I'm trying to say that the operation makes some guarantee that the qubit returned on the nth wire is in fact the same qubit that was taken in on the nth wire. To make this more precise we can say that when executed they will be the same logical qubit?

Reply:
I'm still not sure what we lose by omitting point 2. What do we need this guarantee for?